src/main/java/baritone/process/elytra/NetherPathfinderContext.java

```diff
@@ -24,98 +24,153 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.PalettedContainer;
 import net.minecraft.world.phys.Vec3;
+import sun.misc.Unsafe;
 
 import java.lang.ref.SoftReference;
+import java.lang.reflect.Field;
+import java.nio.file.Path;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import static net.minecraft.world.level.chunk.LevelChunkSection.SECTION_SIZE;
 
 /**
  * @author Brady
  */
 public final class NetherPathfinderContext {
 
+    private static final Unsafe UNSAFE;
+    static {
+        try {
+            Field f = Unsafe.class.getDeclaredField("theUnsafe");
+            f.setAccessible(true);
+            UNSAFE = (Unsafe) f.get(null);
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
     private static final BlockState AIR_BLOCK_STATE = Blocks.AIR.defaultBlockState();
     // This lock must be held while there are active pointers to chunks in java,
     // but we just hold it for the entire tick so we don't have to think much about it.
-    public final Object cullingLock = new Object();
+    public final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
+    public final ReentrantReadWriteLock.ReadLock readLock = rwl.readLock();
+    public final ReentrantReadWriteLock.WriteLock writeLock = rwl.writeLock();
 
     // Visible for access in BlockStateOctreeInterface
     final long context;
     private final long seed;
-    private final ExecutorService executor;
+    // write locked operations
+    private final ExecutorService writeExecutor = Executors.newSingleThreadExecutor();
+    // operations that don't make changes to the chunk cache. could use multiple threads but i'm not sure if it would cause problems.
+    private final ExecutorService readExecutor = Executors.newSingleThreadExecutor();
+    private final ResourceKey<Level> dimension;
 
-    public NetherPathfinderContext(long seed) {
-        this.context = NetherPathfinder.newContext(seed);
+    public NetherPathfinderContext(long seed, Path cache, Level world) {
+        this.dimension = world.dimension();
+        final int dim;
+        if (this.dimension == Level.NETHER) dim = NetherPathfinder.DIMENSION_NETHER;
+        else if (this.dimension == Level.END) dim = NetherPathfinder.DIMENSION_END;
+        else dim = NetherPathfinder.DIMENSION_END;
+        int height = Math.min(world.dimensionType().height(), 384);
+        if (!Baritone.settings().elytraAllowAboveRoof.value && dim == NetherPathfinder.DIMENSION_NETHER) height = Math.min(height, 128);
```

Comment:
Ouch. Can you please autoformat your code with whatever formatting the rest of Baritone uses?

Reply:
using properly formatted braces would make it take up more lines for no benefit